### PR TITLE
Public API documentation

### PR DESCRIPTION
**Pull request by lemnik**
## Goal
Write documentation comments for the public API.

## Testing
Read through generated documentation using `dartdoc`

### CODE REVIEW

File: packages/bugsnag_flutter/lib/src/callbacks.dart [hunk 1]
```diff
@@ -2,6 +2,15 @@ import 'dart:async';
 
 import 'model.dart';
 
+/// A callback to be run before error reports are sent to Bugsnag.
+///
+/// You can use this to add or modify information attached to an error
+/// before it is sent to your dashboard. You can also return `false` from any
+/// callback to halt execution.
+///
+/// "on error" added in Dart are only triggered for events originating
```
Comment (nickdowell) [inline packages/bugsnag_flutter/lib/src/callbacks.dart:11]:
```suggestion
/// "on error" callbacks added in Dart are only triggered for events originating
```

File: packages/bugsnag_flutter/lib/src/model/device.dart [hunk 2]
```diff
@@ -44,10 +73,26 @@ class Device {
...
+/// - [BugsnagEvent]
+/// - [App]
 class DeviceWithState extends Device {
+  /// The number of free bytes of storage available on the device
   int? freeDisk;
+
+  /// The number of free bytes of memory available on the device
   int? freeMemory;
+
+  /// The orientation of the device when the event occurred: either
```
Comment (nickdowell) [inline packages/bugsnag_flutter/lib/src/model/device.dart:91]:
```suggestion
  /// The orientation of the device when the event occurred: e.g.
```

There are more possible values than the two listed 😂

File: packages/bugsnag_flutter/lib/src/model/event.dart [hunk 3]
```diff
@@ -10,29 +10,71 @@ import 'stackframe.dart';
...
   String? context;
+
+  /// The grouping hash of the event to override the default grouping on the
+  /// dashboard. All events with the same grouping hash will be grouped together
+  /// into one error. This is an advanced usage of the library and mis-using it
+  /// will cause your events not to group properly in your dashboard.
+  ///
+  /// As the name implies, this option accepts a hash of sorts.
   String? groupingHash;
   bool _unhandled;
```
Comment (tomlongridge) [inline packages/bugsnag_flutter/lib/src/model/event.dart:45]: Might it be an idea to group the private fields above/below the public ones? Or alongside the fields that they are relevant to?

File: packages/bugsnag_flutter/lib/src/model/event.dart [hunk 4]
```diff
@@ -41,22 +83,51 @@ class BugsnagEvent {
         (_unhandled != _originalUnhandled) ? true : null;
   }
 
+  /// Adds a map of multiple metadata key-value pairs to the specified section.
   void addMetadata(String section, MetadataSection metadata) =>
```
Comment (tomlongridge) [inline packages/bugsnag_flutter/lib/src/model/event.dart:87]: You amended the type `Map<String, Object>` in `leaveBreadcrumb` - is it correct to use `MetadataSection` here?

**Review by tomlongridge — APPROVED** [review]

File: packages/bugsnag_flutter/lib/src/model/device.dart [hunk 5]
```diff
@@ -44,10 +73,26 @@ class Device {
...
+/// - [BugsnagEvent]
+/// - [App]
 class DeviceWithState extends Device {
+  /// The number of free bytes of storage available on the device
   int? freeDisk;
+
+  /// The number of free bytes of memory available on the device
   int? freeMemory;
+
+  /// The orientation of the device when the event occurred: eg.
```
[the remark this answers is not in the document]
Comment (nickdowell) [reply packages/bugsnag_flutter/lib/src/model/device.dart]: These are not the only possible values

[the remark this answers is not in the document]
Comment (nickdowell) [reply packages/bugsnag_flutter/lib/src/model/device.dart]: For example on iOS - https://github.com/bugsnag/bugsnag-cocoa/blob/f3baf6c0f2a0a3b5a28ec02cb366f72ab5620dd2/Bugsnag/Helpers/BSGUtils.m#L24-L30

File: packages/bugsnag_flutter/lib/src/model/device.dart [hunk 6]
```diff
@@ -44,10 +73,26 @@ class Device {
...
+/// - [App]
 class DeviceWithState extends Device {
+  /// The number of free bytes of storage available on the device
   int? freeDisk;
+
+  /// The number of free bytes of memory available on the device
   int? freeMemory;
+
+  /// The orientation of the device when the event occurred: i.e.
+  /// portrait or landscape
```
Comment (nickdowell) [inline packages/bugsnag_flutter/lib/src/model/device.dart:92]: These are not the only two possible values...

```suggestion
  /// The orientation of the device when the event occurred: e.g.
  /// portrait or landscape
```

**Review by nickdowell — APPROVED** [review]

**Pull request merged**